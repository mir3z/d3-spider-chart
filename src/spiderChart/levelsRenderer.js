import * as d3 from "d3";

/**
 * Renders concentric levels inside *parent* group.
 * @param parent SVG parent group
 * @param innerRadius inner radius at which levels start
 * @param outerRadius outer radius at which levels end
 * @param levels Number of levels to render
 * @returns {function}
 */
export default function levelsRenderer(parent, { innerRadius, outerRadius, levels }) {
    const group = parent
        .append("g")
        .attr("class", CLASS_NAME.LEVELS_GROUP);

    const setLevelLine = (x, y, r) => el => el
        .attr("x1", (d, i) => x(i, r))
        .attr("y1", (d, i) => y(i, r))
        .attr("x2", (d, i) => x(i + 1, r))
        .attr("y2", (d, i) => y(i + 1, r));

    const renderLevels = (group, axisList, levelsCount) => {
        const x = (i, r) => r * Math.sin(i * Math.PI * 2 / axisList.length);
        const y = (i, r) => -r * Math.cos(i * Math.PI * 2 / axisList.length);

        const radiusScale = d3.scaleLinear()
            .domain([0, levelsCount])
            .range([innerRadius, outerRadius]);

        const level = group
            .selectAll(`.${ CLASS_NAME.LEVEL }`)
            .data(d3.range(0, levelsCount + 1));

        const enterLevel = level.enter()
            .append("g")
            .attr("class", CLASS_NAME.LEVEL);

        group.selectAll(`.${ CLASS_NAME.LEVEL }`)
            .each((d, i, nodes) => {
                const group = d3.select(nodes[i]);
                const r = radiusScale(i);

                group
                    .selectAll(`.${ CLASS_NAME.LEVEL_LINE }`)
                    .call(setLevelLine(x, y, r));
            });

        enterLevel.each((d, i, nodes) => {
            const group = d3.select(nodes[i]);
            const r = radiusScale(i);

            const levelLine = group
                .selectAll(CLASS_NAME.LEVEL_LINE)
                .data(axisList);

            levelLine.enter()
                .append("line")
                .attr("class", CLASS_NAME.LEVEL_LINE)
                .call(setLevelLine(x, y, r));

            levelLine.exit().remove();
        });
    };

    return function render(data) {
        const axisList = data[0].values
            .map(d => d.axis);

        renderLevels(group, axisList, levels);

        return group;
    };
}

const CLASS_NAME = {
    LEVELS_GROUP: "levels",
    LEVEL: "level",
    LEVEL_LINE: "level-line"
};
