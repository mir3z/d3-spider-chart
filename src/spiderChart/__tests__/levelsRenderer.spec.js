import levelsRenderer from "../levelsRenderer";
import { cleanup, setup } from "./utils";

describe("levels renderer", () => {
    const defaultOptions = {
        innerRadius: 10,
        outerRadius: 100,
        levels: 3
    };

    const data = [{
        values: [
            { axis: "Statement 1" },
            { axis: "Statement 2" },
            { axis: "Statement 3" }
        ]
    }];

    afterEach(cleanup);

    it("renders its own group", () => {
        const svg = setup();
        const renderLevels = levelsRenderer(svg, defaultOptions);

        const group = renderLevels(data);
        const ownGroup = svg.node().querySelector("g.levels");

        expect(ownGroup).not.toBeNull();
        expect(group.node()).toEqual(ownGroup);
    });

    it("renders all levels", () => {
        const svg = setup();
        const renderLevels = levelsRenderer(svg, defaultOptions);

        const group = renderLevels(data);

        expect(group.node()).toMatchSnapshot();
    });
});
