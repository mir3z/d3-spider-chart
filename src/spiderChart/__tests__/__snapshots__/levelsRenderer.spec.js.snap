// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`levels renderer renders all levels 1`] = `
<g
  class="levels"
>
  <g
    class="level"
  >
    <line
      class="level-line"
      x1="0"
      x2="8.660254037844387"
      y1="-10"
      y2="4.999999999999998"
    />
    <line
      class="level-line"
      x1="8.660254037844387"
      x2="-8.660254037844386"
      y1="4.999999999999998"
      y2="5.000000000000004"
    />
    <line
      class="level-line"
      x1="-8.660254037844386"
      x2="-2.4492935982947065e-15"
      y1="5.000000000000004"
      y2="-10"
    />
  </g>
  <g
    class="level"
  >
    <line
      class="level-line"
      x1="0"
      x2="34.64101615137755"
      y1="-40"
      y2="19.999999999999993"
    />
    <line
      class="level-line"
      x1="34.64101615137755"
      x2="-34.64101615137754"
      y1="19.999999999999993"
      y2="20.000000000000018"
    />
    <line
      class="level-line"
      x1="-34.64101615137754"
      x2="-9.797174393178826e-15"
      y1="20.000000000000018"
      y2="-40"
    />
  </g>
  <g
    class="level"
  >
    <line
      class="level-line"
      x1="0"
      x2="60.62177826491071"
      y1="-70"
      y2="34.999999999999986"
    />
    <line
      class="level-line"
      x1="60.62177826491071"
      x2="-60.621778264910695"
      y1="34.999999999999986"
      y2="35.00000000000003"
    />
    <line
      class="level-line"
      x1="-60.621778264910695"
      x2="-1.7145055188062946e-14"
      y1="35.00000000000003"
      y2="-70"
    />
  </g>
  <g
    class="level"
  >
    <line
      class="level-line"
      x1="0"
      x2="86.60254037844388"
      y1="-100"
      y2="49.99999999999998"
    />
    <line
      class="level-line"
      x1="86.60254037844388"
      x2="-86.60254037844385"
      y1="49.99999999999998"
      y2="50.00000000000004"
    />
    <line
      class="level-line"
      x1="-86.60254037844385"
      x2="-2.4492935982947064e-14"
      y1="50.00000000000004"
      y2="-100"
    />
  </g>
</g>
`;
